### ISSUE TITLE
e_signature_algorithm_not_supported should warn not err for RSA PSS

### ISSUE THREAD

**Issue by monomosc** [zmap/zlint#326]
The ```e_signature_algorithm_not_supported``` Lint currently prohibits PSS-signed Certificates. Is there a reason for that?

Specifically, the line

```
if c.SignatureAlgorithm == x509.SHA1WithRSA || c.SignatureAlgorithm == x509.SHA256WithRSA || c.SignatureAlgorithm == x509.SHA384WithRSA || c.SignatureAlgorithm == x509.SHA512WithRSA || c.SignatureAlgorithm == x509.DSAWithSHA1 || c.SignatureAlgorithm == x509.DSAWithSHA256 || c.SignatureAlgorithm == x509.ECDSAWithSHA1 || c.SignatureAlgorithm == x509.ECDSAWithSHA256 || c.SignatureAlgorithm == x509.ECDSAWithSHA384 || c.SignatureAlgorithm == x509.ECDSAWithSHA512 || c {
```
is missing the ```SHA256WithRSAPSS,SHA384WithRSAPSS,SHA512WithRSAPSS``` SignatureAlgorithm OIDs. I couldn't find anything in the BRs prohibiting those Signature Algorithms.

Note that rsaPSS is also sometimes referred to as PKCS#1 2.1.

**Comment by zakird** [comment]
I'm not seeing anything in the BRs that specifically preclude this, at least in 6.1.5.

**Comment by sleevi** [comment]
Yeah, there’s some active discussion around it.

Here’s the status:
1. BRs 6.1.5 are ambiguous as to the potential RSA algorithms that can be used.
2. Currently, Mozilla Firefox, Google Chrome, and Apple Safari do not support PSS certificates, nor does the Chromium-based Edge. I’m not sure for the IE-based Edge
3. Mozilla Policy 2.7 is considering either limiting the encoding or outright forbidding.

It’s definitely worthy of a significant warning, if demoted from error.

**Comment by monomosc** [comment]
Maybe an interesting side note: All browsers except Mozilla do not complain at all when presented a valid Certificate signed with rsaPSS.

And just 4 days ago Mozilla [announced support of PSS](https://github.com/mozilla/pkipolicy/blob/2.7/rootstore/policy.md#511-rsa)
after finally resolving this [age-old issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1088140)

I am actually very sure that rsaPSS is absolutely valid under RFC 5280 and the BRs. If no one else has done it by then, I will prepare a PR for the ```e_signature_algorithm_not_supported``` Lint including Tests by next weekend.

**Comment by sleevi** [comment]
> Maybe an interesting side note: All browsers except Mozilla do not complain at all when presented a valid Certificate signed with rsaPSS.

I don’t believe that’s true. Could you describe how you tested? 

> And just 4 days ago Mozilla [announced support of PSS](https://github.com/mozilla/pkipolicy/blob/2.7/rootstore/policy.md#511-rsa)

I wrote that PR. It wasn’t intended to indicate support, just expectations. The NSS library still has issues here.

**Comment by monomosc** [comment]
> > And just 4 days ago Mozilla announced support of PSS
>
> I wrote that PR. It wasn’t intended to indicate support, just expectations. The NSS library still has issues here.

That's interesting and not at all clear to me. I got it from here: https://bugzilla.mozilla.org/show_bug.cgi?id=1088140, citing directly

> Note that we're planning to explicitly call out RSA-PSS as a valid encoding in the 2.7 update of the Mozilla Root Store policy. I plan to include a note that these encodings are not currently supported by Firefox.

which does not make that clear.

> > Maybe an interesting side note: All browsers except Mozilla do not complain at all when presented a valid Certificate signed with rsaPSS.

> I don’t believe that’s true. Could you describe how you tested?

On it. Give me some time.

**Comment by sleevi** [comment]
Right, there are “layers” of support being discussed.

Of relevance are both support at the SPKI layer (keys expressed with an OID other than rsaEncryption) and at the signature layer. That’s also why I asked about your testing methodology, and Wayne trying to clarify on that bug that PSS SPKIs are not supported (and, related, PSS signatures are not supported by Mozilla::PKIX or by the NSS CERT layers)

**Comment by monomosc** [comment]
So I did absolutely nothing special and:

Chrome:
![chrome](https://user-images.githubusercontent.com/24259919/69074617-05309780-0a30-11ea-8ee2-bd5c2bfd3018.png)

Edge:
![edge](https://user-images.githubusercontent.com/24259919/69074626-0792f180-0a30-11ea-9227-6b7039134926.png)


Note that this is in our company-internal network with company-internally trusted Roots (Root is signed with PKCS#1 1.5, Intermediate and End-Entity are signed with PKCS#1 2.1 (PSS))

I have not done anything special with this browser, it's just whatever version our client-admins decide to throw at us today. I'm too lazy (and it's franklky very hard to get a Mac to connect to that server owing to our network) to try it out with Safari.

**Comment by sleevi** [comment]
Thanks. I omitted Windows because CAPI is the one platform that supports
PSS signatures, and presently, Chrome added support to continue to
interoperate with our Windows users. For a portion of users, that support
is disabled :)

**Comment by cpu** [comment]
> It’s definitely worthy of a significant warning, if demoted from error.

:+1: Seems reasonable to change `e_signature_algorithm_not_supported` to return `lints.Warn` for this case instead of `lints.Error`. 

>  If no one else has done it by then, I will prepare a PR for the e_signature_algorithm_not_supported Lint including Tests by next weekend.

@monomosc I can commit to reviewing a PR that would impl. the change to warning. Thanks!

**Comment by monomosc** [comment]
> Thanks. I omitted Windows because CAPI is the one platform that supports
PSS signatures, and presently, Chrome added support to continue to
interoperate with our Windows users. For a portion of users, that support
is disabled :)

Ooooh, I see. Didn't know that. Damn those windows-loving ******.

Ok then. If my opinion were to count for anything, I'd vote for at least reducing the linter level to Warning, as it might cause compatibility issues for customers of Web PKI, but I could imagine it making sense for a CA giving a customer the option of specifically ordering a PSS-signed Certificate.

Looking at the code again, I suppose, there should be one Error-level lint for Hash Algorithm Compat to BRs , and one Warning-level check for PSS vs PKCS#1 1.5.

**Comment by zakird** [comment]
I agree, let's move this to a warning since that's what's in the BRs. If Mozilla policies prohibits, we can add another lint to match that document and to cite appropriately.

**Comment by cpu** [comment]
> If no one else has done it by then, I will prepare a PR for the e_signature_algorithm_not_supported Lint including Tests by next weekend.

@monomosc Are you still interested in submitting a PR to change this linter's result level for the RSA PSS case? No pressure, just looking to make sure if you don't have the time that it gets picked up by someone else.

**Comment by monomosc** [comment]
Hi, I have started work on this and then got interrupted - I'm still looking to finish this but if I haven't at least submitted a PR soon I understand if someone else picks this up.

**Comment by cpu** [comment]
> Let's see how things look in Jan. If you have a PR ready great, if not I'll write one.

:wave: I ended up writing a quick PR to address this: https://github.com/zmap/zlint/pull/342

Since this issue was opened [v2.7 of the Mozilla root program](https://github.com/mozilla/pkipolicy/pull/199/files) was adopted and like @sleevi mentioned it has language to specifically forbid the RSASSA-PSS signature scheme OID:
> CAs MUST NOT use the id-RSASSA-PSS OID (1.2.840.113549.1.1.10) within a SubjectPublicKeyInfo to represent a RSA key.

Once the open PRs are resolved I'd like to come back around on the new Mozilla root program reqs. and write more specific lints for the conditions outlined in the diff. For now #342 simply updates the existing CABF based lint.

@monomosc PTAL.
